x-pack/plugins/asset_manager/server/lib/implicit_collection/collectors/services.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { APM_INDICES } from '../../../constants';
+import { Asset } from '../../../../common/types_api';
+import { CollectorOptions } from '.';
+
+const MISSING_KEY = '__unknown__';
+
+export async function collectServices({ client, from }: CollectorOptions): Promise<Asset[]> {
+  const dsl = {
+    index: APM_INDICES,
+    size: 0,
+    sort: [
+      {
+        '@timestamp': 'desc',
+      },
+    ],
+    _source: false,
+    query: {
+      bool: {
+        filter: [
+          {
+            range: {
+              '@timestamp': {
+                gte: 'now-1h',
+              },
+            },
+          },
+        ],
+        must: [
+          {
+            exists: {
+              field: 'service.name',
+            },
+          },
+        ],
+      },
+    },
+    aggs: {
+      service_environment: {
+        multi_terms: {
+          size: 100,
+          terms: [
+            {
+              field: 'service.name',
+            },
+            {
+              field: 'service.environment',
+              missing: MISSING_KEY,
+            },
+          ],
+        },
+        aggs: {
+          container_host: {
+            multi_terms: {
+              size: 100,
+              terms: [
+                { field: 'container.id', missing: MISSING_KEY },
+                { field: 'host.hostname', missing: MISSING_KEY },
+              ],
+            },
+          },
+        },
+      },
+    },
+  };
+
+  const esResponse = await client.search(dsl);
+  const serviceEnvironment = esResponse.aggregations?.service_environment as { buckets: any[] };
+
+  const services = (serviceEnvironment?.buckets ?? []).reduce<Asset[]>((acc: Asset[], hit: any) => {
+    const [serviceName, environment] = hit.key;
+    const containerHosts = hit.container_host.buckets;
+
+    const service: Asset = {
+      '@timestamp': new Date().toISOString(),
+      'asset.kind': 'service',
+      'asset.id': serviceName,
+      'asset.ean': `service:${serviceName}`,
+      'asset.references': [],
+      'asset.parents': [],
+    };
+
+    if (environment !== MISSING_KEY) {
+      service['service.environment'] = environment;
+    }
+
+    containerHosts.forEach((nestedHit: any) => {
+      const [containerId, hostname] = nestedHit.key;
+      if (containerId !== MISSING_KEY) {
+        (service['asset.parents'] as string[]).push(`container:${containerId}`);
```

Comment:
I see this is where services can get multi-container parents. Not a problem on its own, but we need to think about what it means to be a service vs service instance/process...